app/models/locations/map.rb

```diff
@@ -0,0 +1,56 @@
+module Locations
+  class Map
+    include ActiveModel::Model
+
+    def initialize(params = {})
+      @@locator_url = "http://library.princeton.edu/locator/index.php?"
+      @@stackmap_url = "http://princeton.stackmap.com/view/?"
+      @params = params
+    end
+
+    def url
+      if @params[:loc] && @params[:id]
+        holding_location = get_holding_location(@params[:loc])
+        bibrec = get_bibrec(@params[:id])
+        if !holding_location.nil? && !bibrec.nil?
+          lib = get_lib(holding_location.locations_library_id)
+          if lib.code == 'firestone'
+            @@locator_url + "loc=" + @params[:loc]+ "&id=" + @params[:id]
+          else
```

Comment:
The old voyager system is hardcoded to provide a link to a map for every item regardless of location. In OL we only show the map link for accessible locations. Given that the else clause here should probably redirect you to something useful for the ID like the request system page for the material. The elseif clause could checks to see if the location is in a stackmapable library. The fallback else can send the user to https://pulsearch.princeton.edu/requests/{bib_id}.